src/Main.java

```diff
@@ -5,7 +5,38 @@ public static void main(String[] args) {
     }
 
     public static boolean balanced(String exp) {
-        return false;
+
+        // create a stack half the size of the input.
+        // we won't need more it would be impossible to balance that many open brackets.
+        IntStack openBrackets = new IntStack(exp.length()/2 + 1);
+
+        for (char c : exp.toCharArray()) {
+
+            // skip characters
+            if(Character.isAlphabetic(c))
+                continue;
+
+            // open brackets are pushed on to the stack
```

Comment:
Another way: when we see an open bracket, push the corresponding close bracket onto the stack. When we see a close bracket, pop the stack and check.

Reply:
Here's a suggestion:

```suggestion
            // open brackets are pushed on to the stack
            if (something)
                        return;
```